backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var snapshots = this.Context.ProjectSnapshots
+                .Include(s => s.Project).ThenInclude(p => p.Agency)
+                .Include(s => s.Project).ThenInclude(p => p.Status)
+                .Where(s => s.SnapshotOn == report.From || s.SnapshotOn == report.To)
+                .ToArray();
+
+            foreach(ProjectSnapshot snapshot in snapshots)
+            {
+                var previousSnapshot = snapshots.Where(s => s.ProjectId == snapshot.ProjectId && s.Id != snapshot.Id).FirstOrDefault();
+                snapshot.BaselineIntegrity = (snapshot?.NetProceeds ?? 0) - (previousSnapshot?.NetProceeds ?? 0);
+            }
+
+            return snapshots;
+        }
+
+        /// <summary>
+        /// Get all project snapshots matching either the 'To' or 'From' dates of the passed report.
+        /// </summary>
+        /// <param name="reportId"></param>
+        /// <exception cref="ArgumentNullException">Argument 'reportId' is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var report = this.Context.ProjectReports
+                .FirstOrDefault(p => p.Id == reportId) ?? throw new KeyNotFoundException();
+
+            return GetSnapshots(report);
+        }
+
+        /// <summary>
+        /// Add the specified report to the datasource.
+        /// The 'From' date is set to the 'To' date of the most recent report, if one exists.
+        /// For every SPL project in the system, a new Snapshot is generated using current project data and a 'snapshotOn' time equal to the newly created reports 'To' date.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to add report.</exception>
+        /// <returns></returns>
+        public ProjectReport Add(ProjectReport report)
+        {
+            if (report.To == null)
+            {
+                report.To = DateTime.UtcNow;
+            }
+            var mostRecentReport = this.Context.ProjectReports
```

Comment:
Because of how the frontend is implemented, the from will never be populated during the add workflow. I can change this if necessary.

Reply:
Ideally the initial report generated would use the `From` from the most recent `IsFinal` report (or the most recent report if there is no final one).  You could determine that before clicking the `Add` button, as you would have a list of reports in memory.

Reply:
My intention was to actually have the backend do what you are recommending in your comment. Does that work?

Reply:
maybe something like this:
```
var mostRecentReport = this.Context.ProjectReports
                .Where(r => r.ReportTypeId == ReportTypes.SPL)
                .OrderByDescending(r => r.IsFinal)
                .ThenByDescending(r => r.To)
                .FirstOrDefault();
```

Reply:
I think it does make sense for the UI to be able to submit the `From` value.  As it would increase the flexibility.  But allow the backend to figure it out if no `From` is provided.

Reply:
That LINQ looks correct.